redeployContract.js

```diff
@@ -134,27 +137,32 @@ const portContract = async (contract_address,
     }
     
 
-    // TODO: Clean up this mess
-    let deploy_code = `0x${createB.createBytecode(contract_code, storage)}`;
-    // TODO: Should calculate proper gas requirement for executing code
-    console.log('Length', deploy_code.length);
-    if (deploy_code.length < code_size) {
-        return await deployLogic(target_web3, target_address, deploy_code);
+    // calculate how much the migration costs
+    const deploy_code = `0x${createB.createBytecode(contract_code, storage)}`;
+    const initContractInstance = new target_web3.eth.Contract([]);
+    const estimatedGas = await initContractInstance.deploy({ data: deploy_code }).estimateGas({
+        to: target_address,
+        gas: config.chain.gasLimit
+    });
+    console.log(`Estimated gas for migration: ${estimatedGas}`);
+
+    if ((estimatedGas * config.chain.gasBufferMultiplicator) <= config.chain.gasLimit) {
```

Comment:
Multiplying here seems quite drastic. I think a constant buffer would be sufficient here:

`if ((estimatedGas + config.chain.gasBuffer) <= config.chain.gasLimit) {`

Reply:
I tried it with a constant buffer, but the actual cost was always unpredictably higher than the estimated one. One time estimatedGas was 35000, real gas was 90000 (2.5x), another time estimatedGas was 65000 and the real gas was 350000 (5.5x). It feels like a constant buffer does not take these unpredictable fluctuations of the estimation into account.

Reply:
after additional testing I noticed that estimatedGas fluctuates depending on the performed action on the blockchain. For example: differences between estimated gas cost and actual gas cost for deploying a new contract fluctuate greatly, while the estimated gas cost for migrating key value pairs mostly hold true.

I would propose to have both, the gasBufferMultiplicator and a gasBuffer and use them for appropriate use cases

Reply:
Have you tried passing the parameters using `encodeABI()`? Seems like that helped for someone having a similar issue: https://ethereum.stackexchange.com/questions/11053/how-to-estimate-gas-at-contract-creation-deployment-to-private-ethereum-blockcha/11114

Reply:
I think the problem is rooted in the contract initialization. Setting a `from` value there seems to solve the problem. At least I get very accurate estamates using the following call:

```
const logicContractInstance = new target_web3.eth.Contract([], undefined,
    {
        from: target_address,
        data: deploy_code
    });
const estimatedGas = await logicContractInstance.deploy().estimateGas();
```

Reply:
For some reason I put 'to' instead of 'from' into the options parameter in order to estimate the gas.
Now it works.